Closes #20. Add `Sentry::Config` for managing configuration, and check `.sentry.yml` on startup

faultyserver opened the pull request:
This PR introduces `Sentry::Config` as a single object to manage the configuration of a `sentry` instance. The CLI also now checks for a `.sentry.yml` file in the project directory to load configurations automatically. These configurations can still be overridden by specifying options the command line, though this is now entirely optional, as all options are supported from the YAML file.

An example and explanation of the `.sentry.yml` format can be found in `.sentry.example.yml`.

faultyserver committed:
Add `Sentry::Config` for managing configuration, and check for `.sentry.yml`.

This commit introduces `Sentry::Config` as a single object to manage the configuration of a `sentry` instance. The CLI also now checks for a `.sentry.yml` file in the project directory to load configurations automatically. These configurations can still be overwritten by specifying options the command line, though this is now entirely optional, as all options are supported from the YAML file.

An example and explanation of the `.sentry.yml` format can be found in `.sentry.example.yml`.

samueleaton commented:
very quick turnaround. Do you think it would be worth allowing users to specify their own path to the config file (e.g. `-c`, `--config`) like I mentioned in the Issue?

And could you run `crystal tool format` on `src` directory? :)

faultyserver commented:
I knew I forgot something! Coming right up.

samueleaton commented on src/sentry.cr:
```diff
@@ -1,6 +1,99 @@
+    end
+
+
+    def to_s(io : IO)
```
I think the config is better aligned when like this:

```ruby
def to_s(io : IO)
  io << <<-CONFIG
  🤖  Sentry configuration:
        process_name: #{name}
        info:         #{info}
        build:        #{build}
        build_args:   #{build_args}
        run:          #{run}
        run_args:     #{run_args}
        watch:        #{watch}
  CONFIG
end
```

And the added robot emoji distinguishes sentry output from the app output. 👍

samueleaton commented on src/sentry_cli.cr:
```diff
@@ -11,66 +11,57 @@ begin
+config = Sentry::Config.from_yaml(config_yaml)
+
+
+
```
wee too many line breaks here

samueleaton commented on src/sentry.cr:
```diff
@@ -1,6 +1,99 @@
+    end
+
+    def build
+      @build ||= "crystal build ./src/#{name}.cr"
```
hmmm.... this bit is a breaking change. Originally, the "process name" was intended to be cosmetic, and only changed what the process was referred to when printed to the terminal. The default build and run commands were based off of the name in the shard.yml file. Now the process name argument will influence the build and run commands. Was that intentional? Maybe the purpose of the process name was unclear or you disagree (which is cool too)?

faultyserver replied:
It was a bit unclear to me. When I originally looked at `sentry_cli.cr`, I saw that `process_name` was being interpolated into the string, so I naively assumed that it was meant to follow what was passed on the CLI, but looking again, that clearly isn't what the behavior was.

That was actually one of the harder things to implement when I was working on this, so I'm kind of glad it's not the intended behavior :) I'll revert that now.

faultyserver committed:
Allow users to specify a config file with `-c` or `--config`

faultyserver committed:
Rename `process_name` to `display_name`. Use name from shard for default build/run commands.

samueleaton commented on src/sentry.cr:
```diff
@@ -1,28 +1,144 @@
-                   @run_args : Array(String) = [] of String,
-                   files = [] of String,
-                   should_build = true)
+      @display_name : String,
```
much better name

faultyserver commented:
I'm not entirely sure about the semantics at the end of `sentry_cli.cr`. It's a little different than before, but I think it's safer now?

Before, if the user specified a name with `--name`, but didn't have a name from `shard.yml` _and_ didn't set a custom build command with `--build`, sentry would start the process runner anyway (because the local `process_name` variable was being re-used).

Now, it will fail if the name isn't set in `shard.yml`, no matter what. Since the display name doesn't affect the default build/run commands, I think this makes sense.

However, if the user sets custom build/run commands, it should be fine that they don't have a name in `shard.yml`.

Not sure how to go about doing that, or if it's worthwhile for what I think would be a very minor edge case.

faultyserver committed:
Only set `display_name` on merge if it was explicitly set on `other`.
The default getter falls back to the shard name, and because the conditional for merging `display_name` only checked that this was truthy, it would always be set to whatever `other` had as a `display_name`, or the shard name (never keeping its own value).

faultyserver committed:
Track whether config explicitly sets `build` and `run` commands.

Similar to `display_name` before, `build` and `run` were also being overwritten when merging configs because of their default values.

This commit adds more boolean properties to `Sentry::Config` to track whether the `build` and `run` commands have been explicitly set, rather than using the default value. These booleans are used in `merge!` to decide whether the config being merged in should apply its `build` and `run` commands to the receiving config.

faultyserver commented:
Quick bump.

I've been using this pretty regularly for the past 2 weeks with no issues, so I think it's pretty much good to go.

Anything else that needs to be changed?

samueleaton commented:
Thanks for the reminder.
